Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -154,90 +288,178 @@ bool FileWatcher::PlatformStart()
     {
         if (QDir(directory).exists())
         {
-            m_platformImpl->AddWatchFolder(directory, recursive);
+            // this happens BEFORE we start the thread that listens to the file queue, so there is no need for a lock here.
+            m_platformImpl->AddWatchFolder(directory, recursive, *this, false);
         }
     }
 
+    AZ_TracePrintf("FileWatcher", "Using %i file watch handles.\n", static_cast<int>(m_platformImpl->m_handleToFolderMap.size()));
+
     return true;
 }
 
 void FileWatcher::PlatformStop()
 {
-    m_shutdownThreadSignal = true;
-
-    m_platformImpl->Finalize();
-
-    if (m_thread.joinable())
+    // close the handle and signal the thread to wake up
+    m_platformImpl->CloseMainWatchHandle();
+  
+    if (m_thread.joinable()) // it may have already finished.
     {
         m_thread.join(); // wait for the thread to finish
     }
-}
 
+    m_platformImpl->Finalize();
+}
 
 void FileWatcher::WatchFolderLoop()
 {
     char eventBuffer[s_inotifyReadBufferSize];
+    
+    int cycleCount = 0;
+
+    constexpr const nfds_t nfds = 2;
+    struct pollfd fds[nfds];
+
+    m_startedSignal = true; // signal that we are no longer going to drop any events.
+
     while (!m_shutdownThreadSignal)
     {
-        ssize_t bytesRead = ::read(m_platformImpl->m_inotifyHandle, eventBuffer, s_inotifyReadBufferSize);
-        if (bytesRead < 0)
+        memset(eventBuffer, 0, s_inotifyReadBufferSize);
+
+        fds[0].fd = m_platformImpl->m_wakeThreadHandle; 
+        fds[0].events = POLLIN;
+        fds[1].fd = m_platformImpl->m_inotifyHandle; 
+        fds[1].events = POLLIN;
+
+        int numPollEvents = poll(fds, nfds, -1);
+        if (numPollEvents == -1) 
+        {
+            break; // error polling.
+        }
+
+        // were we woken up by the wake thread event?
+        if (fds[0].revents & POLLIN)
         {
-            // Break out of the loop when the notify handle was closed (outside of this thread)
             break;
         }
-        if (!bytesRead)
+
+        ssize_t bytesRead = 0;
+        if ((fds[1].revents & POLLIN) && (!m_shutdownThreadSignal))
         {
-            continue;
+            bytesRead = ::read(m_platformImpl->m_inotifyHandle, eventBuffer, s_inotifyReadBufferSize);
+            if (bytesRead < 0)
```

Comment:
Do we need to memset the `eventBuffer` at all?
Since we are going to cast the `&eventBuffer[index]` to an `inotify_event*`, we shouldn't need to zero out the buffer.

Reply:
I did it for pure safety here, as the 'self events' from the OS (as in, 'self was deleted') don't actually write to the name ptr or name buffer at all, so if someone tries to printf it, it printfs fragments of garbage.

Reply:
Ok, so that is why.
Thanks for the explanation.